File: problems/029/029.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    int sum = 0;

    FILE *fp;
    const char* FILE_NAME = "problems/029/temp.dat";
    remove(FILE_NAME);
    fp = fopen(FILE_NAME, "w+");

    for (int i = 2; i <= 100; i++)
    {
        char n_str[200];
        sprintf(&n_str[0], "%02d", i);

        int str_length = 2;

        char products_array[200][200];

        for (int j = 2; j <= 100; j++)
        {
            for (int m = (str_length - 1); m >= 0; m--)
            {
                char temp_product[3+1];
                const char temp_char[2] = { n_str[m] };
                sprintf(temp_product, "%03d", i * atoi(temp_char));

                for (int k1 = 0; k1 < (m + 1); k1++)
                    products_array[m][k1] = '0';

                for (int k2 = (m + 1); k2 < (m + 4); k2++)
                    products_array[m][k2] = temp_product[k2-m-1];

                for (int k3 = (m + 4); k3 < (str_length + 3); k3++)
                    products_array[m][k3] = '0';
            }

            memset(n_str, 0, sizeof(n_str));

            n_str[str_length + 3 - 2] = '0';

            for (int m = (str_length + 3 - 1); m > 0; m--)
            {
                int temp_sum = 0;
                for (int n = 0; n < str_length; n++)
                    temp_sum += products_array[n][m] - '0';

                temp_sum  += n_str[m-1] - '0';
                n_str[m-1] = ((temp_sum % 10) + '0');
                n_str[m-2] = ((int)(temp_sum / 10) + '0');
            }

            str_length += 2;

            char n_str_copy[200];
            memset(n_str_copy, '0', sizeof(n_str_copy));
            n_str_copy[199] = 0;

            memcpy(&n_str_copy[200-str_length], &n_str[0], str_length);

            char line[201];
            int  duplicate_found = 0;
            fseek(fp, 0, SEEK_SET);
            while (fgets(line, sizeof(line), fp) != NULL)
            {
                if (strcmp(line, n_str_copy) == 0)
                {
                    duplicate_found = 1;
                    break;
                }
            }

            if (!duplicate_found)
            {
                fseek(fp, 0, SEEK_END);
                fprintf(fp, "%s", n_str_copy);
            }
        }
    }

    char line[201];
    fseek(fp, 0, SEEK_SET);
    while (fgets(line, sizeof(line), fp))
        sum++;

    fclose(fp);
    remove(FILE_NAME);

    printf("%d\n", sum);

    return 0;
}

File: problems/029/029.f90
PROGRAM F029

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    CHARACTER (LEN=200) :: N_STR, N_STR_COPY
    CHARACTER (LEN=201) :: LINE
    CHARACTER (LEN=3)   :: TEMP_PRODUCT

    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/029/temp.dat"

    CHARACTER, DIMENSION(201,201) :: PRODUCTS_ARRAY

    INTEGER :: SUM_VAL = 0

    INTEGER :: I, J, K1, K2, K3, M, N, TEMP_SUM, STR_LENGTH

    OPEN (2, FILE = FILENAME, STATUS = 'REPLACE', ACTION='READWRITE')

    DO I = 2, 100

        WRITE(N_STR, '(I0.2)') I

        STR_LENGTH = 2

        PRODUCTS_ARRAY = ACHAR(0)

        DO J = 2, 100

            DO M = STR_LENGTH, 1, -1

                WRITE(TEMP_PRODUCT, '(I0.3)') I * (IACHAR(N_STR(M:M)) - IACHAR('0'))

                DO K1 = 1, M
                    PRODUCTS_ARRAY(M, K1) = '0'
                END DO

                DO K2 = (M + 1), (M + 3)
                    PRODUCTS_ARRAY(M, K2) = TEMP_PRODUCT(K2-M:K2-M)
                END DO

                DO K3 = (M + 4), (STR_LENGTH + 3)
                    PRODUCTS_ARRAY(M, K3) = '0'
                END DO

            END DO

            N_STR = REPEAT("0", 200)

            DO M = (STR_LENGTH + 3), 2, -1

                TEMP_SUM = 0
                DO N = 1, STR_LENGTH
                    TEMP_SUM = TEMP_SUM + IACHAR(PRODUCTS_ARRAY(N, M)) - IACHAR('0')
                END DO

                TEMP_SUM = TEMP_SUM + IACHAR(N_STR(M-1:M-1)) - IACHAR('0')
                N_STR(M-1:M-1) = ACHAR(MOD(TEMP_SUM, 10) + IACHAR('0'))
                IF ((M - 2) > 0)                                               &
                N_STR(M-2:M-2) = ACHAR(FLOOR(REAL(TEMP_SUM) / 10.0) + IACHAR('0'))

            END DO
!IF (I == 99 .AND. J == 100) PRINT *, N_STR, NEW_LINE('1')
            STR_LENGTH = STR_LENGTH + 2

            N_STR_COPY = REPEAT("0", 200)

            N_STR_COPY(201-STR_LENGTH:) = N_STR(1:STR_LENGTH)

            REWIND 2
            DO
                READ(2, '(A)', END=10) LINE
                IF (LINE == N_STR_COPY) GOTO 20
            END DO

            10 CONTINUE

            BACKSPACE 2
            WRITE(2, '(A)') N_STR_COPY

            20 CONTINUE

        END DO

    END DO

    REWIND 2
    DO
        READ(2, '(A)', END=30) LINE
        SUM_VAL = SUM_VAL + 1
    END DO

    30 CLOSE(2, STATUS='DELETE')

    PRINT ('(I0)'), SUM_VAL

END PROGRAM F029
